src/lib.rs

```diff
@@ -8,86 +8,612 @@
-    for (name, value) in table {
-        let version = match value {
-            toml::Value::String(ref s) => s,
-            toml::Value::Table(ref t) => {
-                let mut feature = None;
-                let mut version = None;
-                for (tname, tvalue) in t {
-                    match (tname.as_str(), tvalue) {
-                        ("feature", &toml::Value::String(ref s)) => {
-                            feature = Some(s);
+use std::path::{Path, PathBuf};
+use std::str::FromStr;
+use strum_macros::EnumString;
+use thiserror::Error;
+use version_compare::VersionCompare;
+
+/// Metadeps errors
+#[derive(Error, Debug)]
+pub enum Error {
+    /// pkg-config error
+    #[error(transparent)]
+    PkgConfig(#[from] pkg_config::Error),
+    /// One of the `Config::add_build_internal` closures failed
+    #[error("Failed to build {0}: {1}")]
+    BuildInternalClosureError(String, #[source] BuildInternalClosureError),
+    /// Failed to read `Cargo.toml`
+    #[error("{0}")]
+    FailToRead(String, #[source] std::io::Error),
+    /// Raised when an error is detected in the metadata defined in `Cargo.toml`
+    #[error("{0}")]
+    InvalidMetadata(String),
+    /// Raised when dependency defined manually using `METADEPS_$NAME_NO_PKG_CONFIG`
+    /// did not define at least one lib using `METADEPS_$NAME_LIB` or
+    /// `METADEPS_$NAME_LIB_FRAMEWORK`
+    #[error("You should define at least one lib using {} or {}", flag_override_var(.0, "LIB"), flag_override_var(.0, "LIB_FRAMEWORK"))]
+    MissingLib(String),
+    /// An environnement variable in the form of `METADEPS_$NAME_BUILD_INTERNAL`
+    /// contained an invalid value (allowed: `auto`, `always`, `never`)
+    #[error("{0}")]
+    BuildInternalInvalid(String),
+    /// Metadeps has been asked to internally build a lib, through
+    /// `METADEPS_$NAME_BUILD_INTERNAL=always' or `METADEPS_$NAME_BUILD_INTERNAL=auto',
+    /// but not closure has been defined using `Config::add_build_internal` to build
+    /// this lib
+    #[error("Missing build internal closure for {0} (version {1})")]
+    BuildInternalNoClosure(String, String),
+    /// The library which has been build internally does not match the
+    /// required version defined in `Cargo.toml`
+    #[error("Internally built {0} {1} but minimum required version is {2}")]
+    BuildInternalWrongVersion(String, String, String),
+}
+
+#[derive(Error, Debug)]
+/// Error used in return value of `Config::add_build_internal` closures
+pub enum BuildInternalClosureError {
+    /// `pkg-config` error
+    #[error(transparent)]
+    PkgConfig(#[from] pkg_config::Error),
+    /// General failure
+    #[error("{0}")]
+    Failed(String),
+}
+
+impl BuildInternalClosureError {
+    /// Create a new `BuildInternalClosureError::PkgConfig` wrapping
+    /// a `pkg-config` error
+    ///
+    /// # Arguments
+    ///
+    /// * `error`: the `pkg-config` to wrap
+    pub fn pkg_config(error: pkg_config::Error) -> Self {
```

Comment:
Couldn't this just be a `From` impl, and actually `thiserror` generates one for this above already